libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -51,39 +50,58 @@ bool KnowledgeBase::knownToBeDifferent(YulString _a, YulString _b)
 	return false;
 }
 
+optional<u256> KnowledgeBase::differenceIfKnownConstant(YulString _a, YulString _b)
+{
+	// Try to use the simplification rules together with the
+	// current values to turn `sub(_a, _b)` into a constant.
+
+	Expression expr1 = simplify(FunctionCall{{}, {{}, "sub"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
+	if (Literal const* value = get_if<Literal>(&expr1))
+		return valueOfLiteral(*value);
+
+	return {};
+}
+
 bool KnowledgeBase::knownToBeDifferentByAtLeast32(YulString _a, YulString _b)
 {
 	// Try to use the simplification rules together with the
 	// current values to turn `sub(_a, _b)` into a constant whose absolute value is at least 32.
 
-	Expression expr1 = simplify(FunctionCall{{}, {{}, "sub"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
-	if (holds_alternative<Literal>(expr1))
-	{
-		u256 val = valueOfLiteral(std::get<Literal>(expr1));
-		return val >= 32 && val <= u256(0) - 32;
-	}
+	if (optional<u256> difference = differenceIfKnownConstant(_a, _b))
+		return difference >= 32 && difference <= u256(0) - 32;
 
 	return false;
 }
 
+bool KnowledgeBase::knownToBeZero(YulString _a)
+{
+	return valueIfKnownConstant(_a) == u256{};
+}
+
+optional<u256> KnowledgeBase::valueIfKnownConstant(YulString _a)
+{
+	if (m_variableValues.count(_a))
+		if (Literal const* literal = get_if<Literal>(m_variableValues.at(_a).value))
+			return valueOfLiteral(*literal);
+	return {};
+}
+
 Expression KnowledgeBase::simplify(Expression _expression)
 {
-	bool startedRecursion = (m_recursionCounter == 0);
-	ScopeGuard{[&] { if (startedRecursion) m_recursionCounter = 0; }};
+	return simplifyRecursively(move(_expression), 0);
+}
 
-	if (startedRecursion)
-		m_recursionCounter = 100;
-	else if (m_recursionCounter == 1)
+Expression KnowledgeBase::simplifyRecursively(Expression _expression, size_t _recursionCounter)
+{
+	if (_recursionCounter > 100)
 		return _expression;
-	else
-		--m_recursionCounter;
 
 	if (holds_alternative<FunctionCall>(_expression))
 		for (Expression& arg: std::get<FunctionCall>(_expression).arguments)
-			arg = simplify(arg);
+			arg = simplifyRecursively(arg, _recursionCounter + 1);
 
 	if (auto match = SimplificationRules::findFirstMatch(_expression, m_dialect, m_variableValues))
```

Comment:
Maybe the reason for the knowledge base needing the additional rule is that this needs to be a ``while``?

Reply:
there is a recursive call: https://github.com/ethereum/solidity/pull/11817/files#diff-689e75ac518209a32f0ffa72f01340bebfa32b36ac25f5ff1cc17b99ed876b0eR104